frontend/lib/resolution/resolution-queries.cpp

```diff
@@ -1559,6 +1534,15 @@ const TypedFnSignature* instantiateSignature(Context* context,
           useType = getInstantiationType(context,
                                          actualType,
                                          formalType);
+
+          // Verify that the 'instantiation type' still accepts the actual.
+          // This might not be the case based on legal argument mapping rules.
+          auto kind = resolveIntent(useType, /* isThis */ false, /* isInit */ false);
+          auto useTypeConcrete = QualifiedType(kind, useType.type(), useType.param());
+
+          if (!canPass(context, actualType, useTypeConcrete).passes()) {
```

Comment:
Why is this necessary?

Reply:
See the PR description. In particular, we don't validate instantiations that also require conversions for certain intents. E.g., if instantiating `ref x: Parent`, we allow it to be instantiated with `shared Child`, but `ref x: shared Parent` does not accept `shared Child`, since the `ref` intent requires the types to be equal.